cmd/test-server/main.go

```diff
@@ -178,48 +196,164 @@ func pingHandler(w http.ResponseWriter, r *http.Request) {
+	icmp := RtItem{ip, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt), fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)}
+	return icmp
+}
 
+func TcpPinger(ip string) tcpStruct {
 	// TCP Pinger
 	var tcpResultArr []tcpResult
 	rand.Seed(time.Now().UnixNano()) // Or each time we restart server the sequences would repeat
+	var portsWaitGroup sync.WaitGroup
+	portsWaitGroup.Add(len(PortsToTest))
 	for _, port := range PortsToTest {
-		var seqNumber uint64 = uint64(rand.Uint32())
-		var dst = fmt.Sprintf("%s:%d", clientIP, port)
-		ticker := time.NewTicker(TCPInterval)
-		var tResult []float64
-		for x := 0; x < TCPCounter; x++ {
-			seqNumber++
-			select {
-			case <-ticker.C:
-				tResult = append(tResult, pingTcp(dst, seqNumber, TCPTimeout))
+		go func(port int) {
+			defer portsWaitGroup.Done()
+			var seqNumber uint64 = uint64(rand.Uint32())
+			var dst = fmt.Sprintf("%s:%d", ip, port)
+			ticker := time.NewTicker(TCPInterval)
+			var tResult []float64
+			for x := 0; x < TCPCounter; x++ {
+				seqNumber++
+				select {
+				case <-ticker.C:
+					tResult = append(tResult, sendTcpPing(dst, seqNumber, TCPTimeout))
+				}
+			}
+			ticker.Stop()
+			min, avg, max, stddev := getStats(tResult)
+			tcpResultArr = append(tcpResultArr, tcpResult{port, tResult, min, avg, max, stddev})
+		}(port)
+	}
+	portsWaitGroup.Wait()
+	tcpResultObj := tcpStruct{ip, tcpResultArr}
+	return tcpResultObj
+}
+
+// Avg RTT from all successful ICMP measurements, to display on webpage
+func getAvgIcmpResult(icmp []RtItem) float64 {
+	var sum float64 = 0
+	var len float64 = 0
+	for _, x := range icmp {
+		if x.AvgRtt == 0 {
+			continue
+		}
+		sum += x.AvgRtt
+		len += 1
+	}
+	var avg float64 = sum / len
+	return avg
+}
+
+// Avg RTT from all successful TCP measurements regardless of port, to display on webpage
+func getAvgTcpResult(tcp []tcpStruct) float64 {
+	var sum float64 = 0
+	var len float64 = 0
+	// for each IP in subnet
+	for _, x := range tcp {
+		// for each port per IP
+		for _, p := range x.Probes {
+			if p.AvgRtt == 0 {
+				continue
 			}
+			sum += p.AvgRtt
+			len += 1
+		}
+	}
+	var avg float64 = sum / len
+	return avg
+}
+
+// Handler for ICMP and TCP measurements which also serves the webpage via a template
+func pingHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/ping" {
+		http.NotFound(w, r)
+		return
+	}
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusNotImplemented)
+		w.Write([]byte(http.StatusText(http.StatusNotImplemented)))
+		return
+	}
+	clientIPstr := r.RemoteAddr
+	clientIP, _, _ := net.SplitHostPort(clientIPstr)
+	adjIPstoPing := getAdjacentIPs(clientIP)
+	var expUUID = uuid.NewString()
+	var timestamp = time.Now().Format("2006-01-02T15:04:05.000000") //RFC3339 style date with added seconds information
```

Comment:
I like to use `time.Now().UTC()` to guarantee that the time is in UTC. Otherwise, you're asking for trouble with time zone math if the system's clock isn't set to UTC. Also, like `expUUID`, you could do this directly in the struct below.